src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+                // We could consider looking at the reason for failure and
+                // punishing the peer, but for now just give up on sync.
+                Finalize();
+                return ret;
+            }
+        }
+
+        ret.success = true;
+        // Return any headers that are ready for acceptance.
+        ret.headers_to_process = RemoveHeadersReadyForAcceptance();
+
+        // If we hit our target blockhash, then all remaining headers will be
+        // returned and we can clear any leftover internal state.
+        if (m_redownloaded_headers.empty() && m_process_all_remaining_headers) {
+            Finalize();
+            return ret;
+        }
+
+        // If the headers message is full, we need to request more.
+        if (full_headers_message) {
+            ret.locator = MakeNextHeadersRequest();
+            return ret;
+        } else {
+            // For some reason our peer gave us a high-work chain, but is now
+            // declining to serve us that full chain again. Give up.
+            // Note that there's no more processing to be done with these
+            // headers, so we can still return success.
+            Finalize();
+            return ret;
+        }
+    }
+    return ret;
+}
+
+bool HeadersSyncState::ValidateAndStoreHeadersCommitments(const std::vector<CBlockHeader>& headers)
+{
+    // The caller should not give us an empty set of headers.
+    Assume(headers.size() > 0);
+    if (headers.size() == 0) return true;
+
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return false;
+
+    if (headers[0].hashPrevBlock != m_last_header_received.GetHash()) {
+        // Somehow our peer gave us a header that doesn't connect.
+        // This might be benign -- perhaps we issued an extra getheaders
+        // message, such as after a block INV was received.
+        // Or it could be that our peer is broken or malicious. If broken,
+        // sending a new getheaders immediately could trigger an infinite
+        // loop. Just give up for now; if our peer ever gives us an block
+        // INV later we will fetch headers then, and likely retrigger this
+        // logic.
+        Finalize();
+        return false;
+    }
+
+    // If it does connect, (minimally) validate and store a commitment to each one.
+    for (const auto& hdr : headers) {
+        if (!ValidateAndProcessSingleHeader(m_last_header_received, hdr, m_current_height+1)) {
+            Finalize();
+            return false;
+        }
+    }
+
+    if (m_current_chain_work >= m_minimum_required_work) {
+        m_blockhash_with_sufficient_work = headers.back().GetHash();
+        m_redownloaded_headers.clear();
+        m_redownload_buffer_last_height = m_chain_start->nHeight;
+        m_redownload_buffer_first_prev_hash = m_chain_start->GetBlockHash();
+        m_redownload_buffer_last_hash = m_chain_start->GetBlockHash();
+        m_download_state = State::REDOWNLOAD;
```

Comment:
5691c174c5cb2f9947e6cc2db8d32b4b4715e47d: it would be good to have a log message when the state switches to `REDOWNLOAD`.

Reply:
Done.